File: sujioh/2024-07-12-path-sum-ii.py
class Solution:
    def pathSum(self, root: TreeNode, targetSum: int) -> List[List[int]]:
        def dfs(root, targetSum, path):
            if not root:
                return []
            targetSum -= root.val
            path.append(root.val)
            if not root.left and not root.right:  # Is leaf node
                if targetSum == 0:  # Found a valid path
                    ans.append(path.copy())
            else:
                dfs(root.left, targetSum, path)
                dfs(root.right, targetSum, path)
            path.pop()  # backtrack

        ans = []
        dfs(root, targetSum, [])
        return ans

File: jun/2024-07-12-path-sum-ii.ts
const dfs = (root: TreeNode | null, targetSum: number, temp: number[], result: number[][]) => {
    if (!root) return
    temp.push(root.val)
    const val = targetSum - root.val
    if (!root.left && !root.right) {
        if (val === 0) {
            result.push(temp)
        }
        return
    }
    dfs(root.left, val, [...temp], result)
    dfs(root.right, val, [...temp], result)
}

function pathSum(root: TreeNode | null, targetSum: number): number[][] {
    const result: number[][] = []

    dfs(root, targetSum, [], result)
    return result
}
